server/resp/src/main/java/org/infinispan/server/resp/commands/json/JSONSET.java

```diff
@@ -0,0 +1,139 @@
+   static Configuration config = Configuration.builder().options(Option.DEFAULT_PATH_LEAF_TO_NULL)
+         .jsonProvider(new com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider())
+         .mappingProvider(new com.jayway.jsonpath.spi.mapper.JacksonMappingProvider())
+         .build();
+
+   public static void jsonSetBiConsumer(CharSequence value, ResponseWriter writer) {
+      if (value == null) {
+         // Use null consumer
+         writer.nulls();
+         return;
+      }
+      if (value.length() == 2 && value.charAt(0) == 'O' && value.charAt(1) == 'K') {
+         // Use ok consumer
+         writer.ok();
+         return;
+      }
+      // All the rest are errors
+      writer.error(value);
+   }
+
+   public JSONSET() {
+      super("JSON.SET", -3, 1, 1, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx,
+         List<byte[]> arguments) {
+      byte[] key = arguments.get(0);
+      byte[] path = arguments.get(1);
+      byte[] value = arguments.get(2);
+      if (arguments.size() > 4) {
+         handler.writer().syntaxError();
+         return handler.myStage();
+      }
+      boolean nx, xx; // must be effectively final
+      if (arguments.size() == 4) {
+         String arg = (new String(arguments.get(3)).toUpperCase());
+         switch (arg) {
+            case NX:
+               nx = true;
+               xx = false;
+               break;
+            case XX:
+               xx = true;
+               nx = false;
+               break;
+            default:
+               handler.writer().syntaxError();
+               return handler.myStage();
+         }
+      } else {
+         nx = false;
+         xx = false;
+      }
+      ObjectMapper mapper = new ObjectMapper();
+         JsonNode newNode;
+         try {
+            newNode = mapper.readTree(value);
+         } catch (IOException e) {
+            handler.writer().error(e.getMessage());
+            return handler.myStage();
+         }
+         FunctionalMap.ReadWriteMap<byte[], Object> cache = ReadWriteMapImpl
+               .create(FunctionalMapImpl.create(handler.typedCache(null)));
+         CompletionStage<CharSequence> cs = cache.eval(key, view -> {
+            var doc = (byte[]) view.find().orElse(null);
+            if (doc == null) {
+               if (xx) {
+                  return null;
+               }
+               if (!isRoot(path)) {
+                  return "-ERR new objects must be created at root";
+               }
+               view.set(value);
+               return RespConstants.OK;
+            }
+            if (nx) {
+               return null;
+            }
+            try {
+               var rootObjectNode = (ObjectNode) mapper.readTree(doc);
+               var jpCtx = JsonPath.using(config).parse(rootObjectNode);
+               var pathStr = new String(path);
+               jpCtx.set(pathStr, newNode);
```

Comment:
I see we only verify the flags against the entry. We should verify them against the documents before applying the change.

```text
127.0.0.1:6379> JSON.SET test $ '{"a": 1, "b": 2}'
OK
127.0.0.1:6379> JSON.SET test "$.c" 3 NX
OK
127.0.0.1:6379> JSON.SET test "$.d" 4 XX
(nil)
```